vendor/github.com/minio/minio-go/bucket-cache.go

```diff
@@ -0,0 +1,159 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage (C) 2015 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package minio
+
+import (
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"sync"
+)
+
+// bucketRegionCache provides simple mechansim to hold bucket regions in memory.
+type bucketRegionCache struct {
+	// Mutex is used for handling the concurrent
+	// read/write requests for cache
+	sync.RWMutex
+
+	// items holds the cached regions.
+	items map[string]string
+}
+
+// newBucketRegionCache provides a new bucket region cache to be used
+// internally with the client object.
+func newBucketRegionCache() *bucketRegionCache {
+	return &bucketRegionCache{
+		items: make(map[string]string),
+	}
+}
+
+// Get returns a value of a given key if it exists
+func (r *bucketRegionCache) Get(bucketName string) (region string, ok bool) {
+	r.RLock()
+	defer r.RUnlock()
+	region, ok = r.items[bucketName]
+	return
+}
+
+// Set will persist a value to the cache
+func (r *bucketRegionCache) Set(bucketName string, region string) {
+	r.Lock()
+	defer r.Unlock()
+	r.items[bucketName] = region
+}
+
+// Delete deletes a bucket name.
+func (r *bucketRegionCache) Delete(bucketName string) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.items, bucketName)
+}
+
+// getRegion - get region for the bucketName from region map cache.
+func (a API) getRegion(bucketName string) (string, error) {
```

Comment:
getRegion and getBucketLocation should me merged and modularized differently. I will explain on chat.